### PR TITLE
[Server] codegen-server support for restXml (S3)

### PR DESCRIPTION
**Pull request by guymguym**
Signed-off-by: Guy Margalit <guymguym@gmail.com>

## Motivation and Context

See #903

The motivation is to implement the server protocol code for the s3d-rs crate - still draft code in the project but the idea is to provide an S3 gateway for edge devices that handles disconnections and caching transparently. See: https://github.com/s3d-rs/s3d 

## Description

This PR adds generated server code for restXml protocol types as mentioned by @crisidev in the related issue.

Please keep in mind that I am quite new to this project and to kotlin, and I would be happy to get your review and fix anything that you may suggest.

## Testing

Added the S3 model namespace to the "blocklist" in ServerProtocolTestGenerator.kt because the server protocol is not fully implemented.

## Checklist

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by rcoh** [comment]
you may want to install `pre-commit`, then run `pre-commit install` to pick up our pre-commit hooks (including Kotlin autoformatters)

**Comment by guymguym** [comment]
thanks @rcoh - updated the PR with pre-commit

File: rust-runtime/aws-smithy-http-server/src/protocols.rs
Comment (guymguym) [inline rust-runtime/aws-smithy-http-server/src/protocols.rs:11]: I wasn't sure what to do here - I changed it to `check_content_type` and allow json or xml, but the error is `MissingJsonContentType`. Should I change it somehow?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (guymguym) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:243]: Is there a better way I should use to write the `"Error"`tag or it makes sense to hard-quote it here in the code template?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (guymguym) [inline codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:228]: I have no idea if this is correct

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt
Comment (guymguym) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:873]: I had to fix that, but I hope this is also ok for client codegen too

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt
Comment (guymguym) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt:212]: I am not sure yet what is needed here so any directions would be useful

File: rust-runtime/aws-smithy-http-server/src/protocols.rs
Comment (crisidev) [reply rust-runtime/aws-smithy-http-server/src/protocols.rs:11]: The original idea was to have multiple `check_<type>_content_type()` functions and to abstract the call to it inside the protocol implementation.

We are at a very early stage of development for the server codegen and we are working to finish the support of [restJson1](https://awslabs.github.io/smithy/1.0/spec/aws/aws-restjson1-protocol.html). Because of this part of the codebase is not as generic as I would like it to be.

What can be done is to have a switch matching the protocol type inside `ServerHttpProtocolGenerator` and depending on the protocol it will produce the right `check_<type>_content_type()` identifier:

```kotlin
fun getContentTypeCheck(): String {
      when (codegenContext.protocol) {
        RestJson1Trait.ID -> {
            return "check_json_content_type"
        }
        RestXmlTrait.ID -> {
            return "check_xml_content_type"
        }
        else -> {
            TODO("Protocol ${codegenContext.protocol} not supported yet")
        }
    }
}    
```

Later on

```kotlin
...
    val contentTypeCheck = getContentTypeCheck()
    rustTemplate(
        """
           let body = request.take_body().ok_or(#{SmithyHttpServer}::rejection::BodyAlreadyExtracted)?;
           let bytes = #{Hyper}::body::to_bytes(body).await?;
           if !bytes.is_empty() {
               #{SmithyHttpServer}::protocols::${contenTypeCheck}(request)?;
               input = #{parser}(bytes.as_ref(), input)?;
           }
        """,
        *codegenScope,
        "parser" to parser,
    )
...
```

I am suggesting this to unblock your PR. I have opened an issue #987 to further extract this logic inside the Protocol logic that can be addressed later.

Notes: check_json_content_type() is called [here](https://github.com/awslabs/smithy-rs/blob/main/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt#L522)

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (crisidev) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:243]: Probably someone working on the client codegen can answer to this..

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt
Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:873]: I am not sure about the implications of this, I will have to check with the folks working on client codegen.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt
Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt:212]: Since we are at an early phase of development, we do not implement all Smithy protocol traits yet and we want this to hard fail because if a model requires functionalities that are not implemented, it should fail when it run.

**Comment by crisidev** [comment]
The CI is failing. It is important that `./gradlew :codegen-server-test:test` succeeds in your local build.

It is likely some of the protocol tests we generate are not working for restXml or for the S3 model you are using. I suggest you to add the S3 model namespace to the blocklist defined in the protocol tests. You can add another check to [the if condition](https://github.com/awslabs/smithy-rs/blob/6325c8a277225e786920d42df935371d27b11fc4/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt#L153-L158). This is, again, because we are still working on finishing the implementation of our first protocol 😃

**Review by crisidev — CHANGES_REQUESTED** [review]
This is wonderful work! Thanks a million. I left a couple of comments and hopefully explanations to your questions and to how to move forward.

**Comment by guymguym** [comment]
Hi @crisidev 
So I would like to ask for some help with the current failures in the test, below is what I got locally.
The first seem simple that we don't get the <?xml tag. The other are because the bucket parsing doesn't work for some reason. Any ideas of where to look would help a lot. Thanks

```
failures:

---- operation::server_get_bucket_location_test::get_bucket_location_unwrapped_output_response stdout ----
thread 'operation::server_get_bucket_location_test::get_bucket_location_unwrapped_output_response' panicked at 'assertion failed: `(left == right)`
  left: `"<?xml version=\"1.0\" encoding=\"UTF-8\"?>\n<LocationConstraint xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\">us-west-2</LocationConstraint>"`,
 right: `b"<LocationConstraint xmlns=\"http://s3.amazonaws.com/doc/2006-03-01/\"></LocationConstraint>"`', s3/rust-server-codegen/src/operation.rs:85:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- operation::server_list_objects_v2_test::s3_virtual_host_dualstack_accelerate_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_dualstack_accelerate_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:413:9

---- operation::server_list_objects_v2_test::s3_virtual_host_dualstack_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_dualstack_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:333:9

---- operation::server_list_objects_v2_test::s3_default_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_default_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:215:9

---- operation::server_list_objects_v2_test::s3_operation_addressing_preferred_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_operation_addressing_preferred_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:453:9

---- operation::server_list_objects_v2_test::s3_virtual_host_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:254:9

---- operation::server_list_objects_v2_test::s3_virtual_host_accelerate_addressing_request stdout ----
thread 'operation::server_list_objects_v2_test::s3_virtual_host_accelerate_addressing_request' panicked at 'assertion failed: `(left == right)`
  left: `ListObjectsV2Input { bucket: None, delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`,
 right: `ListObjectsV2Input { bucket: Some("mybucket"), delimiter: None, encoding_type: None, max_keys: 0, prefix: None, continuation_token: None, fetch_owner: false, start_after: None, request_payer: None, expected_bucket_owner: None }`', s3/rust-server-codegen/src/operation.rs:373:9


failures:
    operation::server_get_bucket_location_test::get_bucket_location_unwrapped_output_response
    operation::server_list_objects_v2_test::s3_default_addressing_request
    operation::server_list_objects_v2_test::s3_operation_addressing_preferred_request
    operation::server_list_objects_v2_test::s3_virtual_host_accelerate_addressing_request
    operation::server_list_objects_v2_test::s3_virtual_host_addressing_request
    operation::server_list_objects_v2_test::s3_virtual_host_dualstack_accelerate_addressing_request
    operation::server_list_objects_v2_test::s3_virtual_host_dualstack_addressing_request

test result: FAILED. 1 passed; 7 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s

```

File: rust-runtime/aws-smithy-http-server/src/protocols.rs
Comment (guymguym) [reply rust-runtime/aws-smithy-http-server/src/protocols.rs]: I just added this to the PR - I hope this makes sense..

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt
Comment (guymguym) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt]: done.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt
Comment (guymguym) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:876]: For context this change was needed to set the value in operation_deser for primitive types like bool/i32/etc 
For example this generated code:
```
                input = input.set_fetch_owner(
                    crate::operation_deser::parse_str_list_objects_v2_input_fetch_owner(v)?,
                );
```
did not compile as the setter expected an Option<bool>.
I think this is actually a must have change anyhow because the generated body always returns Ok(Some(value)) and just the return type of the function was not matching.
See here - https://github.com/awslabs/smithy-rs/blob/6325c8a277225e786920d42df935371d27b11fc4/codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt#L883

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (guymguym) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:243]: Should we tag someone to ask for review?

**Comment by guymguym** [comment]
Oh ok so after I added the S3 model namespace to the blocklist defined in the protocol tests, the tests do not fail.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt
Comment (guymguym) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt]: Ok so this is also not failing now that I added s3 to the block-list as suggested.

Comment (guymguym) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt]: So I reverted my change here.

**Comment by guymguym** [comment]
@crisidev I addressed the comments and the tests pass now locally.
There are still a couple of open questions on the changes that might affect the client codegen, that I could use your help to tag someone to review.
Thanks!

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt
Comment (crisidev) [inline codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt:155]: Please also update the comment.

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt
Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:876]: @rcoh can you help us here?

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (crisidev) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:243]: @rcoh do you mind to help us here?

**Review by crisidev — APPROVED** [review]
LGTM. I will work on some refactoring after this is merged to close #987 and #988.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (rcoh) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:243]: hardcoding `Error` appears to be correct here: https://awslabs.github.io/smithy/1.0/spec/aws/aws-restxml-protocol.html#error-response-serialization.

Note that you'll also need to handle the conditional error wrapping here. On the client side, we handle it by injecting a different error extractor but on the server side, I think you'll need to handle it slightly differently. The most parallel way would be to have a function that created an `ErrorWriter` that could be swapped as a runtime type, but I'll leave that to @crisidev to decide.

https://github.com/awslabs/smithy-rs/blob/1fa0ecfccef7e73a2886937cc1bb7935c4bf4ec0/codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/RestXml.kt#L63-L66

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt
Comment (guymguym) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/generators/protocol/ServerProtocolTestGenerator.kt]: done

**Review by rcoh — COMMENTED** [review]
overall changes look good to me! The real work will be getting all the protocol tests happy. Note that S3 generates an extended error id, I'm not sure if this is required for protocol compliance or not: https://github.com/awslabs/smithy-rs/blob/1aa59693eed10713dec0f3774a8a25ca271dbf39/aws/rust-runtime/aws-inlineable/src/s3_errors.rs#L32

**Comment by rcoh** [comment]
(and for your own context, the diff and doc preview tools unfortunately rely on this being a PR directly to the repo which is why they're failing We hopefully will get those setup as something standalone at some point)

**Review by rcoh — APPROVED** [review]
changes to `codegen` look fine, delegating final approval to @crisidev

File: codegen-server-test/build.gradle.kts
Comment (rcoh) [inline codegen-server-test/build.gradle.kts:28]: probably not in scope for this PR, but we should add the restXml protocol tests here:https://github.com/awslabs/smithy-rs/blob/cf46613360c184c5a45b7299858ebc72b9b71013/codegen-test/build.gradle.kts#L35-L38

but I suspect it might be a bit rough until we iron out all of the issues 😅

File: codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt
Comment (crisidev) [reply codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpProtocolGenerator.kt:876]: I think we can go with makeOptional() as this stage. Input validation is on the road map and it is something we will address later.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (crisidev) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:243]: In the server there is no need for the injection. We handle an error coming from here as a [serialization rejections](https://github.com/awslabs/smithy-rs/blob/main/rust-runtime/aws-smithy-http-server/src/rejection.rs#L29-L34).

File: codegen-server-test/build.gradle.kts
Comment (crisidev) [reply codegen-server-test/build.gradle.kts:28]: It in plans. We will probably work on this next week.

File: codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt
Comment (guymguym) [reply codegen/src/main/kotlin/software/amazon/smithy/rust/codegen/smithy/protocols/serialize/XmlBindingTraitSerializerGenerator.kt:228]: thanks, reesolving

**Comment by guymguym** [comment]
should I rebase?

**Comment by crisidev** [comment]
Rebase done, let's wait for the CI.

**Comment by crisidev** [comment]
I am pretty sure there will still be a lot of things to tweak in both restJson and RestXml, but this is a great start! Thanks @guymguym for the help, we really appreciate it.

**Comment by guymguym** [comment]
no problem @crisidev I will be happy to look for more opportunities to contribute

**Comment by guymguym** [comment]
And thank you for the super fast responses and helpful approach. Same goes to @rcoh.

**Pull request merged**
